Fix for issue #11505

gmantri opened the pull request:
Fixes the issue #11505 (https://github.com/Azure/azure-sdk-for-js/issues/11505)

joheredi commented on sdk/storage/storage-file-share/src/ShareClient.ts:
```diff
@@ -988,12 +988,22 @@ export class ShareClient extends StorageClient {
       };
 
       for (const identifier of response) {
+        let accessPolicy: any = undefined;
```
Is there a Type you can use here instead of any?

gmantri replied:
Thanks for the question. Please see [here](https://docs.microsoft.com/en-us/rest/api/storageservices/set-share-acl#remarks). From the **Establishing Share-Level Access Policies** section

> A stored access policy can specify the start time, expiry time, and permissions for the shared access signatures with which it is associated. Depending on how you want to control access to your share or file resource, you can specify all of these parameters within the stored access policy, and omit them from the URL for the shared access signature. Doing so permits you to modify the associated signature's behavior at any time, as well as to revoke it. Or you can specify one or more of the access policy parameters within the stored access policy, and the others on the URL. Finally, you can specify all of the parameters on the URL.

Based on this (and my past experience working with Storage REST API), only thing that is required when defining an access policy is the identifier. All other parameters are optional.

Regarding your 2nd question about using a Type instead of any, I simply copied the code from Storage Queue SDK (https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/storage/storage-queue/src/QueueClient.ts#L934) as there the implementation is correct :)

ljian3377 replied:
It's a pity we don't have a proper interface defined for this.
```ts
{
    /**
     * @member {Date} startsOn the date-time the policy is active.
     */
    startsOn?: Date;
    /**
     * @member {string} expiresOn the date-time the policy expires.
     */
    expiresOn?: Date;
    /**
     * @member {string} permissions the permissions for the acl policy
     * @see https://docs.microsoft.com/en-us/rest/api/storageservices/set-share-acl
     */
    permissions: string;
  };
```

The  `AccessPolicy ` from underlying use `string` for `Date`.
```ts
export interface AccessPolicy {
  /**
   * The date-time the policy is active.
   * **NOTE: This entity will be treated as a string instead of a Date because the API can
   * potentially deal with a higher precision value than what is supported by JavaScript.**
   */
  startsOn?: string;
  /**
   * The date-time the policy expires.
   * **NOTE: This entity will be treated as a string instead of a Date because the API can
   * potentially deal with a higher precision value than what is supported by JavaScript.**
   */
  expiresOn?: string;
  /**
   * The permissions for the ACL policy.
   */
  permissions?: string;
}
```

joheredi commented on sdk/storage/storage-file-share/src/ShareClient.ts:
```diff
@@ -988,12 +988,22 @@ export class ShareClient extends StorageClient {
-            permissions: identifier.accessPolicy!.permissions!,
-            startsOn: new Date(identifier.accessPolicy!.startsOn!)
-          },
+          accessPolicy,
```
What does it mean to set an ACL with only the identifier but no access policy? Is it a valid scenario?

/cc: @XiaoningLiu @jeremymeng

ljian3377 replied:
It's valid since fileds in the access policy are used in combined with these in the SAS parameters. Though it makes this access policy of no use though. SDK won't add any special check for this as long as the service thinks it's OK.

joheredi replied:
Thanks @gmantri and @ljian3377 for the details! I think it makes sense!

joheredi reviewed (commented):
Thanks a lot for your contribution! The changes look good in general, I just have a question about the scenario where an ACL is set without an access policy as the REST API [documentation](https://docs.microsoft.com/en-us/rest/api/storageservices/set-share-acl#request-body) doesn’t seem to explicitly say which elements are required.

ljian3377 commented:
The change itself looks good to me.

Two additional issues:

- `accessPolicy` in `SignedIdentifier` should have its fields as optional.

https://github.com/Azure/azure-sdk-for-js/blob/7f97d0ef44d388b092e27f21d3e49c98975287ed/sdk/storage/storage-file-share/src/ShareClient.ts#L233

```ts
export interface SignedIdentifier {
  /**
   * @member {string} id a unique id
   */
  id: string;
  /**
   * @member {AccessPolicy} accessPolicy
   */
  accessPolicy: {
    /**
     * @member {Date} startsOn the date-time the policy is active.
     */
    startsOn: Date;
    /**
     * @member {string} expiresOn the date-time the policy expires.
     */
    expiresOn: Date;
    /**
     * @member {string} permissions the permissions for the acl policy
     * @see https://docs.microsoft.com/en-us/rest/api/storageservices/set-share-acl
     */
    permissions: string;
  };
}
```
-  `ShareClient.setAccessPolicy()`.

https://github.com/Azure/azure-sdk-for-js/blob/7f97d0ef44d388b092e27f21d3e49c98975287ed/sdk/storage/storage-file-share/src/ShareClient.ts#L1038

In File:
```ts
      for (const identifier of shareAcl || []) {
        acl.push({
          accessPolicy: {
            expiresOn: truncatedISO8061Date(identifier.accessPolicy.expiresOn),
            permissions: identifier.accessPolicy.permissions,
            startsOn: truncatedISO8061Date(identifier.accessPolicy.startsOn)
          },
          id: identifier.id
        });
      }
```
In Queue
```ts
      for (const identifier of queueAcl || []) {
        acl.push({
          accessPolicy: {
            expiresOn: identifier.accessPolicy.expiresOn
              ? truncatedISO8061Date(identifier.accessPolicy.expiresOn)
              : undefined,
            permissions: identifier.accessPolicy.permissions,
            startsOn: identifier.accessPolicy.startsOn
              ? truncatedISO8061Date(identifier.accessPolicy.startsOn)
              : undefined
          },
          id: identifier.id
        });
      }
```

ljian3377 commented:
Also want to hold the merge of this PR till after our October release as we changed the structure of files and deleted this file in our feature branch. Merging this now would make it difficult to pick up the changes when merging.

gmantri commented:
> Also want to hold the merge of this PR till after our October release as we changed the file structure and deleted this file in our feature branch. Merging this now would make it difficult to pick up the changes when merging.

@ljian3377 ... What's the ETA for the October release? This is a blocking issue for us. One thing you could do is simply take the edits from my PR and make the changes in the feature branch accordingly.

ljian3377 commented:
> @ljian3377 ... What's the ETA for the October release? This is a blocking issue for us. One thing you could do is simply take the edits from my PR and make the changes in the feature branch accordingly.

Around October 10th. You need us to release a pacakge, right? If that's the case, we should consider including this change for our next release. But it's gonna be a prerelease version (preview for new storage features). Is that OK with you?

If prerelease version or October 10th don't work for you, we may need to think about other workarounds.

ljian3377 commented:
Can't change the accessPolicy in SignedIdentifier's fields and itself to optional as it's a breaking change.
Returning undefined for required field will introduce inconsistency but will live with for now till we have a new major release.

ljian3377 commented:
The change is taken and shipped in @azure/storage-file-share@12.3.0-beta.1 with another PR.
